### PR TITLE
[Attention][CPU] Run Zen CPU encoder attention on zentorch SDPA

### PR DESCRIPTION
**Pull request by Priyjain-amd**
## Purpose

On CPU, encoder / encoder-only attention runs a two-step dance: a reshape-and-cache prologue (`_C::cpu_attn_reshape_and_cache`) stages K/V into a scratch cache, then `_C::cpu_attention_with_kv_cache` reads it straight back on the **same** step — even though an encoder never reuses that cache across steps.

`zentorch::zentorch_sdpa` attends the packed QKV **in place**, so this PR routes encoder / encoder-only attention to it on AMD Zen CPUs when both hold:

- the op is registered (zentorch is built with the SDPA op), **and**
- the model dtype clears zentorch's ISA gate.

Guardrails:

- **Batch shaping** — equal-length batches fold into a single dense call for the whole scheduler batch; ragged batches loop per sequence.

This removes the K/V staging round-trip and runs a faster kernel, concentrating the win on the attention op.

## Changes

Three files, **+254** lines; behavior outside the CPU encoder-attention path is untouched.

- **`vllm/v1/attention/backends/zentorch_sdpa.py`** *(new, +122)* — the encoder-SDPA path, two helpers:
  - `should_use_zentorch_sdpa(attn_type, alibi_slopes, sliding_window, dtype)` — the gate. Returns `True` only for `ENCODER` / `ENCODER_ONLY` layers with **no ALiBi and no sliding window**, when the op is registered (`has_zentorch_op(["zentorch_sdpa"])`) and the dtype clears the ISA gate: `bfloat16` → `torch.cpu._is_avx512_bf16_supported()`, `float32` → `torch.cpu._is_avx512_supported()`; `float16` stays native (no runtime query for AVX512-FP16).
  - `zentorch_encoder_sdpa(query, key, value, output, attn_metadata, scale)` — attends the packed QKV in place (`is_causal=False`, no bias) via `torch.ops.zentorch.zentorch_sdpa.out`, writing straight into `output`. Per-sequence lengths come from `attn_metadata.query_start_loc`; **equal-length** batches reshape to a dense `[B, H, S, D]` and run in **one** call for the whole scheduler batch, **ragged** batches loop per sequence. GQA/MQA needs no expansion — the op derives the KV head count from `key`.
- **`vllm/v1/attention/backends/cpu_attn.py`** *(+32)* — wires the gate into `CPUAttentionBackendImpl`: `self.use_zentorch_sdpa` is computed once in `__init__`, and `forward()` short-circuits encoder attention to `zentorch_encoder_sdpa(...)` and returns **before** the scratch-cache staging + `cpu_attention_with_kv_cache` path.
- **`tests/kernels/attention/test_zen_cpu_encoder_sdpa.py`** *(new, +100)* — numerics vs the `ref_varlen_encoder_attn` reference across uniform + ragged batches, GQA `(8, 2)` and MHA `(4, 4)`, head size `{64, 128}`, `{bfloat16, float32}`; plus a truth table for `should_use_zentorch_sdpa` (ALiBi / sliding-window / decoder / encoder-decoder / `float16` all stay native). Skips on non-Zen CPUs or when the op isn't built.

## Op-level performance

`cpu_attention_with_kv_cache` (baseline) vs `zentorch_sdpa` (this PR). Torch-profiler summary — worker rank 0, `record_shapes=True`; same server and workload, only the attention op differs. Both legs captured an identical **12,312 attention calls across 513 forward passes**.

| Metric | `cpu_attention_with_kv_cache` (baseline) | `zentorch_sdpa` (this PR) | Δ |
|---|:--:|:--:|:--:|
| **Attention kernel — self CPU** | **20.527 s** | **12.808 s** | **−37.6% (1.60×)** |
| Attention kernel — avg / call | 1.667 ms | 1.044 ms | −37.4% |
| Attention kernel — calls | 12,312 | 12,312 | — |
| Attention kernel — % of forward | 42.3% | 30.8% | −11.5 pp |
| Attention subgraph — `unified_attention_with_output` | 21.685 s | 14.269 s | −34.2% (1.52×) |
| **Whole forward — self CPU total** | **48.476 s** | **41.542 s** | **−14.3% (1.17×)** |

**Takeaways**

- The kernel itself is **~1.6× faster** (20.53 s → 12.81 s) and the `cpu_attn_reshape_and_cache` staging op (0.86 s) disappears entirely — `zentorch_sdpa` reads packed QKV directly.
- The new path adds ~0.5 s of ATen bookkeeping the staged path avoids: `aten::equal` (~0.44 s, the per-call equal-length batch check) plus small mask/shape ops (`tril`/`triu`/`log`, ~0.1 s). Net attention cost still drops sharply.
- Attention falls from **42.3% → 30.8%** of forward self-CPU time, cutting the whole profiled forward by **14.3%**.

## Test plan

| Setting | Value |
|---|---|
| **Model** | `RedHatAI/embeddinggemma-300m`, pooling runner (MEAN pooling) |
| **Serve args** | `--dtype bfloat16 --max-model-len 2048 --no-enable-prefix-caching` |
| **Workload** | `vllm bench serve --backend openai-embeddings` (random), `--random-input-len 1024 --num-prompts 1024 --max-concurrency 32` |
| **Host** | AMD Zen CPU (`ZenCpuPlatform`), 32 OMP threads pinned to cores 0–31 (`VLLM_CPU_OMP_THREADS_BIND=0-31`) |
| **Stack** | vLLM 0.27.0 · torch 2.13.0+cpu · zentorch 2.13.0.0 |

---

Signed-off-by: priyansh jain <priyansh.jain2@amd.com>

Change-Id: Iab38bd47ef84697c884e4bdb6da217da150d3d38

### CODE REVIEW

**Review by amukho — APPROVED** [review]
Looks good to me...

File: vllm/v1/attention/backends/cpu_attn.py [hunk 1]
```diff
@@ -334,6 +338,21 @@ def __init__(
                 "heads in the layer"
             )
 
+        vllm_config = get_current_vllm_config()
```
[the remark this answers is not in the document]
Comment (Priyjain-amd) [reply vllm/v1/attention/backends/cpu_attn.py:343]: Added.

File: tests/kernels/attention/test_cpu_attn.py [hunk 2]
```diff
@@ -756,6 +778,97 @@ def test_varlen_encoder_attention_amx(
...
+        use_alibi=use_alibi,
+    )
+
+
+@_skip_no_zentorch_sdpa
+@pytest.mark.parametrize(
+    "attn_type,dtype,expected",
+    [
+        (AttentionType.ENCODER_ONLY, torch.bfloat16, True),
+        (AttentionType.ENCODER, torch.float32, True),
```
[the remark this answers is not in the document]
Comment (Priyjain-amd) [reply tests/kernels/attention/test_cpu_attn.py]: Agreed. _skip_no_zentorch_sdpa only covers Zen CPU + op registration, while should_use_zentorch_sdpa also requires AVX-512 BF16 for bfloat16 and AVX-512 for float32.

File: vllm/v1/attention/backends/zentorch_sdpa.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+    Token i attends [i-(W-1), i+(W-1)], matching native cpu_attn and the
+    zentorch plugin (left = right = W-1). Entries inside the band are 0;
+    outside they are -inf. Rank is 4 because zentorch_sdpa rejects 3D masks.
+    """
+    band = window - 1
+    mask = torch.ones((seq_len, seq_len), dtype=dtype)
+    mask = torch.tril(mask, diagonal=band)
+    mask = torch.triu(mask, diagonal=-band)
+    return torch.log(mask).view(1, 1, seq_len, seq_len)
```
[the remark this answers is not in the document]
Comment (Priyjain-amd) [reply vllm/v1/attention/backends/zentorch_sdpa.py]: Agreed.

[the remark this answers is not in the document]
Comment (Priyjain-amd) [reply vllm/v1/attention/backends/zentorch_sdpa.py]: The issue has been fixed, and Zentorch now handles this internally.

File: tests/kernels/attention/test_cpu_attn.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Priyjain-amd) [reply tests/kernels/attention/test_cpu_attn.py]: This is already covered. `_ZENTORCH_SDPA_DTYPES` skips bfloat16 when AVX-512 BF16 isn’t there, and float32 when AVX-512 isn’t there — same checks `should_use_zentorch_sdpa` uses. Both tests pick dtypes from that list, so a Zen box missing either ISA just skips that dtype instead of failing.

File: vllm/v1/attention/backends/zentorch_sdpa.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+if TYPE_CHECKING:
+    from vllm.v1.attention.backends.cpu_attn import CPUAttentionMetadata
+
+__all__ = ["should_use_zentorch_sdpa", "zentorch_sdpa_attn"]
+
+
+def should_use_zentorch_sdpa(
+    attn_type: str,
+    alibi_slopes: torch.Tensor | None,
```
Comment (amukho) [inline vllm/v1/attention/backends/zentorch_sdpa.py]: Is this passing linting since both `alibi_slopes` and `sliding_window` are unused?

File: tests/kernels/attention/test_cpu_attn.py [hunk 5]
```diff
@@ -771,6 +802,116 @@ def test_varlen_encoder_attention_amx(
...
+@pytest.mark.parametrize("sliding_window", [None, -1, 256])
+@pytest.mark.parametrize("dtype", _ZENTORCH_SDPA_DTYPES)
+def test_should_use_zentorch_sdpa(
+    attn_type: str,
+    use_alibi: bool,
+    sliding_window: int | None,
+    dtype: torch.dtype,
+) -> None:
+    alibi_slopes = torch.tensor([1.0]) if use_alibi else None
+    assert should_use_zentorch_sdpa(attn_type, alibi_slopes, sliding_window, dtype)
```
Comment (amukho) [inline tests/kernels/attention/test_cpu_attn.py]: Will this test pass?

**Review by amukho — CHANGES_REQUESTED** [review]
Added some observations...

File: vllm/v1/attention/backends/cpu_attn.py [hunk 6]
```diff
@@ -334,6 +340,21 @@ def __init__(
...
+        self.isa = _get_attn_isa(
```
Comment (dllehr-amd) [inline vllm/v1/attention/backends/cpu_attn.py]: Can we see if we can use the self.isa from the cpu metadata instead of a separate call here

File: vllm/v1/attention/backends/cpu_attn.py [hunk 7]
```diff
@@ -383,6 +389,21 @@ def forward(
             AttentionType.ENCODER,
         )
         if is_encoder_attention:
+            if self.use_zentorch_sdpa and has_zentorch_op(["zentorch_sdpa_attn"]):
```
Comment (amukho) [inline vllm/v1/attention/backends/cpu_attn.py]: Should we fold the op check in the class attribute itself? In that case we can avoid this runtime check

**Review by amukho — CHANGES_REQUESTED** [review]
Added an additional observation...

File: vllm/v1/attention/backends/cpu_attn.py (hunk 6 above)
Comment (Priyjain-amd) [reply vllm/v1/attention/backends/cpu_attn.py]: Yes, We can used that one.

File: vllm/v1/attention/backends/zentorch_sdpa.py (hunk 4 above)
Comment (Priyjain-amd) [reply vllm/v1/attention/backends/zentorch_sdpa.py]: Yes, unused function arguments are not currently enforced by vLLM's lint rules. However, it's generally a good practice to avoid keeping unused arguments in the code. So, removed unused args.

File: vllm/v1/attention/backends/cpu_attn.py (hunk 7 above)
Comment (Priyjain-amd) [reply vllm/v1/attention/backends/cpu_attn.py]: Good catch. The helper already checks that the op is registered when we set `self.use_zentorch_sdpa`, so the extra check in `forward` was redundant. I removed it.

File: tests/kernels/attention/test_cpu_attn.py (hunk 5 above)
Comment (Priyjain-amd) [reply tests/kernels/attention/test_cpu_attn.py]: Yes, it passed.

**Review by amukho — APPROVED** [review]
Looks good to me...

**Review by dllehr-amd — APPROVED** [review]
Looks good.  Thanks @Priyjain-amd   I think I got all of my concerns addressed!
